mutating-webhook/utils.go

```diff
@@ -0,0 +1,213 @@
+					JSONTreePath:   jsonPath,
+					AnnotationPath: annotationPath,
+					FunctionType:   ImportValue,
+				}
+				*needResolving = append(*needResolving, needResolve)
+				stringStack.Pop()
+			} else if hasLabelFunc {
+				start := len("Fn::AddLabel(")
+				end := strings.LastIndex(val, ")")
+				args := strings.Split(val[start:end], ",")
+				fmt.Printf("args: %s %s\n", args[0], args[1])
+
+				keyValLabel := strings.TrimSpace(args[0])
+				fmt.Printf("LABELLLL%s\n", keyValLabel)
+				namespace, kind, crdKindName, subKind, err := ParseCompositionPath(args[1])
+				fmt.Printf("parsed: %s %s %s %s\n", namespace, kind, crdKindName, subKind)
+				jsonData := QueryAPIServer(kind, namespace, crdKindName)
+				fmt.Printf("Results: %s\n", string(jsonData))
+				if err != nil {
+					return err
+				}
+				name, err := ParseDiscoveryJSON(jsonData, subKind)
+				fmt.Println("****************")
+				fmt.Println(name)
+				fmt.Println(err)
+				fmt.Println("****************")
+
+				// listOptions := metav1.ListOptions{
+				// 	FieldSelector: fmt.Sprintf("metadata.name=%s", name),
+				// }
+				// switch subKind {
+				// case "Deployment":
+				// 	deployment := kubeClient.AppsV1().Deployments(namespace)
+				// }
+				needResolve := ResolveData{
+					JSONTreePath:   jsonPath,
+					AnnotationPath: name,
+					FunctionType:   AddLabel,
+				}
+				*needResolving = append(*needResolving, needResolve)
+				stringStack.Pop()
+			} else {
+				stringStack.Pop()
+			}
+		}
+		return nil
+	})
+}
+
+// [Namespace]?.[Kind].[InstanceName].[outputVariable]
+func ParseAnnotationPath(annotationPath string) (string, string, string, string, error) {
+	path := strings.Split(strings.TrimSpace(annotationPath), ".")
+	if len(path) == 3 {
+		return "default", path[0], path[1], path[2], nil
+	}
+	if len(path) != 4 {
+		return "", "", "", "", fmt.Errorf("AnnotationPath inside function is not of len 3 or 4")
+	}
+	return path[0], path[1], path[2], path[3], nil
+}
+
+// [namespace]?.[Kind].[CrdKindName].[SubKind]
+func ParseCompositionPath(compositionPath string) (string, string, string, string, error) {
+	path := strings.Split(strings.TrimSpace(compositionPath), ".")
+	if len(path) == 3 { //can be len 3, namespace opttional
+		return "default", path[0], path[1], path[2], nil
+	}
+	if len(path) != 4 {
+		return "", "", "", "", fmt.Errorf("CompositionPath inside function is not of len 3")
+	}
+	return path[0], path[1], path[2], path[3], nil
+}
+
+func ParseDiscoveryJSON(composition []byte, subKind string) (string, error) {
+	var subName string
+	var err error
+	jsonparser.ArrayEach(composition, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
+		subName, err = ParseDiscoveryJSONHelper(composition, subKind, "")
+	})
+	return subName, err
+}
+
+// Finds the name of the subresource of a CRD resource
+func ParseDiscoveryJSONHelper(composition []byte, subKind string, subName string) (string, error) {
+	if len(subName) > 0 {
+		return subName, nil
```

Comment:
In line 140 name is set. But here in line 127 we are returning subName. Where do we set subName?

Reply:
Please recheck the code . Now it should make more sense, this was from last wedn when it wasn't quite working. i am thinking of changing the current code to returning a string instead of []string btw, that will probably be the last change to this portion of the code